Added C regex to prevent ambiguity in header file

jweinst1 opened the pull request:
Adding a C regex for disambiguation of `.h` files

## Description

Basically, my C header file is getting incorrectly classified as Objective-C. While overrides can fix this, it would be good to have a regex detecting C-only standard header includes. 

https://github.com/jweinst1/DataBlock/blob/master/src/DataBlock.h

That is the url to the incorrectly categorized file.

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] **I am associating a language with a new file extension.**
  - [ ] The new extension is used in hundreds of repositories on GitHub.com
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      - https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3AFOOBAR+KEYWORDS+NOT+nothack
  - [ ] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am adding a new language.**
  - [ ] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
      -  https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3AFOOBAR+KEYWORDS+NOT+nothack
  - [ ] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s):
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [ ] I have included a syntax highlighting grammar.
  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [x] **I am fixing a misclassified language**
  - [ ] I have included a new sample for the misclassified language:
    - Sample source(s): https://github.com/jweinst1/DataBlock/blob/master/src/DataBlock.h
      - [URL to each sample source, if applicable]
    - Sample license(s):
  - [x] I have included a change to the heuristics to distinguish my language from others using the same extension.

- [ ] **I am changing the source of a syntax highlighting grammar**
  <!-- Update the Lightshow URLs below to show the new and old grammars in action. -->
  - Old: https://github-lightshow.herokuapp.com/
  - New: https://github-lightshow.herokuapp.com/

- [ ] **I am adding new or changing current functionality**
  <!-- This includes modifying the vendor, documentation, and generated lists. -->
  - [ ] I have added or updated the tests for the new or changed functionality.

jweinst1 commented:
Yea, I was confused since I thought that you had to do `#include <cstdio>` in C++ but I was mistaken. Ill change this with some truly C only features

jweinst1 commented:
So, following @Elronnd, suggestion on #1626, I changed the C regex to detect `(void)` signified functions, 

Additionally since the [restrict](https://en.wikipedia.org/wiki/Restrict) is only valid in C, I added that to the regex, in addition to [flexible array parameters](https://en.wikipedia.org/wiki/Flexible_array_member)

pchaigno commented on lib/linguist/heuristics.rb:
```diff
@@ -75,6 +75,11 @@ def call(data)
+    CRegex = Regexp.union(
+        /^(char|double|short|int|long|float|unsigned|void)*\w\s*\w\s*\(\s*void\s*\)/,
+        /^(\w|\s|\*)*restrict/,
+        /^(char|short|int|long|float|unsigned)*\s+\w+\s*\[\]\s*;/
```
Why the asterisk after `(char|short|int|long|float|unsigned)`?

jweinst1 replied:
good catch

pchaigno replied:
I think there's still the same error with the two above regular expressions, or am I mistaken?

jweinst1 replied:
No, because u need a return type for any void function, like ` foo(void)` with no return type is not valid.

for the second one i think it's optional to have a word character or a `*` pointer type, since I want to match just the restrict keyword.

pchaigno replied:
> No, because u need a return type for any void function, like ` foo(void)` with no return type is not valid.

Sure, but currently, you can have several return types because of the asterisk. For example, `chard d(void)` matches your regular expression. Also, I think you should require a space after the type and probably meant to write `\w+` instead of just `\w`. Not sure why you have `\w\s*` twice after the type.

> for the second one i think it's optional to have a word character or a `*` pointer type, since I want to match just the restrict keyword.

Could you give an example of C construction you're trying to match with the second regular expression?

jweinst1 replied:
> Sure, but currently, you can have several return types because of the asterisk. For example, doublechard d(void) matches your regular expression. Also, I think you should require a space after the type and probably meant to write \w+ instead of just \w. Not sure why you have \w\s* twice after the type.

Alright, will fix this.

> Could you give an example of C construction you're trying to match with the second regular expression?

Would be:

```c
void copy(void* restrict dst, void* restrict src){...}
```

The restrict keyword is always used with pointers, but most always used in function parameters.

pchaigno replied:
Maybe `\([\s\w\*]+restrict\s+` for the second expression then?

jweinst1 commented:
@pchaigno let me know if this needs any more changes :)

jweinst1 commented:
alright, made your most recent revisions.

jweinst1 commented:
@pchaigno or @lildude , let me know if this PR needs more revisions :)

jweinst1 commented:
@pchaigno feel free to let me know if i need to make any more changes on this.

jweinst1 commented on lib/linguist/heuristics.rb:
```diff
@@ -75,6 +75,11 @@ def call(data)
+    CRegex = Regexp.union(
+        /^(char|double|short|int|long|float|unsigned|void)\w+\s*\(\s*void\s*\)/,
+        /\([\s\w\*]+restrict\s+/,
+        /^(char|short|int|long|float|unsigned)\s+\w+\s*\[\]\s*;/
```
@pchaigno , yes, for the first and the third, the word boundary is more helpful as it will also match common c-types like `unsigned char`, `unsigned int`, `long long` or `unsigned long` made the changes

lildude commented:
As this is the most recently active of attempts to resolve this, I'm going to reference https://github.com/github/linguist/pull/3585 as work that has also started to try and address this. Maybe we can use ideas from both that PR and this to get a good solution in place.

jweinst1 commented:
@pchaigno is this ready to merge? it says only those with write access can merge :(

pchaigno commented:
> @pchaigno is this ready to merge?

No. We still need to wait for the :+1: from someone from GitHub's staff.

pchaigno commented on lib/linguist/heuristics.rb:
```diff
@@ -75,6 +75,11 @@ def call(data)
+    CRegex = Regexp.union(
+        /\b(char|double|short|int|long|float|unsigned|void)\s+\w+\s*\(\s*void\s*\)/,
+        /\([\s\w\*]+restrict\s+/,
+        /\b(char|short|int|long|float|unsigned)\s+\w+\s*\[\]\s*;/
```
What makes you think this last heuristic rule is only valid in C?

jweinst1 replied:
@pchaigno Flexible struct parameters are not valid C++, see https://stackoverflow.com/a/4413035/5129504

pchaigno replied:
`.h` is also [an Objective-C extension](https://github.com/github/linguist/blob/5363e045bb2728eec05ff2225a87e742e0bff586/lib/linguist/languages.yml#L3067).

lildude replied:
… and precisely why the tests are currently failing, suggesting further refinement is needed.

jweinst1 commented:
Alright so i fixed the test failure there, i think the issue was there is a seldom used keyword in Objective-C called `__restrict`, so i removed the underscores from the regex.

jweinst1 commented:
@pchaigno or @lildude , if a keyword is removed from a newer version of a language, does this count as a linguist standard as not being part of the language? 

The [register keyword is removed in C++17](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4340) and has been deprecated for a while before that, yet still exists in C. Can linguist consider register as a C only regex?

lildude commented:
> @pchaigno or @lildude , if a keyword is removed from a newer version of a language, does this count as a linguist standard as not being part of the language?
>
> The register keyword is removed in C++17 and has been deprecated for a while before that, yet still exists in C. Can linguist consider register as a C only regex?

I've just test the spec change you've mentioned and I don't think we can:

> The purpose in removing the usage of register is to preserve the ability to use this valuable keyword for future proposals, such as modules. It is therefore retained as a reserved keyword, with a note on intent, similar to the handling of export in C++11. All other references to the register keyword are removed.

As I understand this, it's no longer used but is still reserved for future use so we'd probably end up having to reverse the change in future if this is resurrected. Probably best to leave this keyword alone for the moment.

jweinst1 commented:
@lildude understood, any other changes you suggest making to this PR? since the tests are passing now

jweinst1 commented:
@pchaigno , @lildude , any other changes need to be made here?

pchaigno commented:
@jweinst1 See my above review :wink: 

> You'll also need to update the tests in `tests/test_heuristics.rb`.

lildude commented:
Nudge.

pchaigno commented:
Flagging as stale.
